### ISSUE TITLE
Add support for Adafruit HT16K33 based LED displays

### ISSUE THREAD

**Issue by patfelst** [esphome/feature-requests#625]
**Describe the problem you have/What new integration you would like**
Could you please add support for Adafruit [LED backpack displays](https://learn.adafruit.com/adafruit-led-backpack/featured_products) all based on the HT16K33 I2C chip. Their products include an LED bargraph, LED 8x8 dot matrix (single and multi-color), 7-segment LED, and 4-digit alpha-numeric LED display.

**Please describe your use case for this integration and alternatives you've tried:**
ESPHome currently does not have support for an alpha-numeric LED display which I am most interested in, yes ESPHome has the max7219 platform, and the upcoming TM1637 but they only support 7-segment displays.

**Additional context**
There is an existing [Arduino library](https://github.com/adafruit/Adafruit_LED_Backpack) which supports all the LED backpack types of displays. It is I2C based so should be a relatively simple adaption of the lcd_pcf8574 display platform.

**Comment by patfelst** [comment]
Samuel Sieb has kindly created a custom component which currently supports just the 4-digit alpha-numeric display. I will not be able to test until the weekend.
https://github.com/ssieb/custom_components/tree/master/lcd_ht16k33

**Comment by docteurzoidberg** [comment]
@ssieb made a custom component for theses, here: https://github.com/ssieb/custom_components/tree/master/ht16k33_alpha
Would be great to include !

**Comment by 42Network** [comment]
How do we vote? I vote for this one for sure. ssieb has quite a few components which could be pulled in to ESPhome.

**Comment by ssieb** [comment]
Pulling isn't something that happens.  I need to write some docs and make PRs.

**Comment by 42Network** [comment]
Yes, I know you'd have to submit a PR. Meanwhile I'm glad for your work and I can include locally. However, I just now realized your code will not work for my HT16K33-based numeric LED display. It actually does work, but randomly displays garbage. I am using it for a simple temperature display, e.g. 104.1. 

I got it working though using this example: https://github.com/JAAlperin/esphome-adafruit-7-segment-LED-backpack-example. This isn't as nice as yours, though, because it requires much more code inside the display lambda.

I am not much of a programmer but I may try to either make your code work with the simple 4 digit numeric display, or may that other code work directly with ESPhome, with the appropriate Python wrappers.

**Comment by ssieb** [comment]
Right, my component is for the alphanumeric displays that have 14 segments, not the 7 segment displays.
Is it the one with the colon in the middle? I just checked the wiring for that one and it seems a little wacky, but a simple component should be easy.  I don't see much point in including the scrolling support for a display like that.

**Comment by 42Network** [comment]
I agree - scrolling for number-only 4-digit 7-segment isn’t needed. But brightness control and access to the dots and colon and possibly individual segment control (ie there are a few letters and symbols that can be shown with just 7 segments) would be useful.

**Comment by cwross** [comment]
Very interested in this as well, particularly for the 4-digit 7-segment display. I can test if needed. Thanks!

**Comment by feaver12** [comment]
i am also interested in support for this same display!!!

**Comment by smartroad** [comment]
Just to throw my towel into the ring, this is also used for Adafruits "Trellis" button/led board (https://www.adafruit.com/product/1616)

Could make a nice HMI self lighting buttons? (showing when things are on/off, maybe flashing for attention?). I don't know enough to add this myself :(

**Comment by cTurtle98** [comment]
any progress on this one? i would love to be able to use this

**Comment by kj4dxk** [comment]
I would like to use it to make a thermostat display.

**Comment by SteveRMann** [comment]
Also interested. I have devices that use the 14-segment display from Adafruit, but I have to run them on Arduino and use MQTT to update them.

**Comment by eifachtimon** [comment]
I have the same Adafruit 7 Segment Display! Is there any progress so far on this one?

**Comment by SteveRMann** [comment]
I did it in ESPHome.  Here is my YAML
Hope it helps.. (I would like to see what you do with it).

`# cgm-monitor.yaml
# This configuration displays the time and glucose value on a
# 4-digit, 14-segment LED on an Adafruit ht16k33 backpack.

esphome:
  name: cgm-monitor
  friendly_name: CGM Monitor

  on_boot:
    priority: -100.0
    then:
      - script.execute: ht16k33_display_startup

esp8266:
  board: d1_mini

# Enable logging
logger:

# Enable I2C on Wemos D1 Mini
i2c:
  id: i2c_bus
  sda: D2
  scl: D1
  scan: true

# Enable Home Assistant API
api:

ota:
  - platform: esphome

#----------------------------- WiFi --------------------------------------
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password


text_sensor:
  # Get the WiFi details
  - platform: wifi_info
    ip_address:
      name: IP
    ssid:
      name: SSID
    mac_address:
      name: Mac Address  

  # Get the Home Assistant Glucose value
  - platform: homeassistant
    id: glucose_sensor
    entity_id: sensor.dexcom_stevermann_glucose_value
    on_value:
      then:
        - script.execute: ht16k33_display_glucose      

  # Get the Home Assistant trend value
  - platform: homeassistant
    id: trend_sensor
    entity_id: sensor.dexcom_trend_val
    internal: true
  
    
#-------------------------------------------------------------------------

web_server:
  port: 80


# Enable time synchronization
time:
  - platform: homeassistant
    id: homeassistant_time
    on_time:
      - seconds: /10
        then:
          - script.execute: ht16k33_display_time
          - lambda: |-
              auto time = id(homeassistant_time).now();
              ESP_LOGI("Time", "Current time: %02d:%02d", time.hour, time.minute);


# Sun component to track sunrise and sunset
sun:
  latitude: !secret latitude
  longitude: !secret longitude

sensor:
  - platform: sun
    name: "Sun Elevation"
    type: elevation
    id: sun_elevation

interval:
  - interval: 60s
    then:
      - if:
          condition:
            lambda: 'return id(sun_elevation).state < 0;'
          then:
            - script.execute: set_brightness_low
          else:
            - script.execute: set_brightness_medium
  - interval: 5s  # More frequent updates for glucose and trend display
    then:
      - script.execute: ht16k33_display_glucose



#########################################################################
########################## HT16K33 Initialization #######################
# Define the HT16K33 Display Initialization and Control
script:
  - id: ht16k33_init
    mode: queued
    then:
      - lambda: |-
         // Time display
         uint8_t init1[] = {0x21}; // Turn on oscillator
          id(i2c_bus).write(0x71, init1, sizeof(init1));
          delay(10);

          uint8_t init2[] = {0x81}; // Display on, no blink
          id(i2c_bus).write(0x71, init2, sizeof(init2));
          delay(10);

          uint8_t init3[] = {0xEF}; // Full brightness
          id(i2c_bus).write(0x71, init3, sizeof(init3));
          delay(10);

         // BG display
         uint8_t init4[] = {0x21}; // Turn on oscillator
          id(i2c_bus).write(0x70, init4, sizeof(init1));
          delay(10);

          uint8_t init5[] = {0x81}; // Display on, no blink
          id(i2c_bus).write(0x70, init5, sizeof(init2));
          delay(10);

          uint8_t init6[] = {0xEF}; // Full brightness
          id(i2c_bus).write(0x70, init6, sizeof(init3));
          delay(10);

  - id: ht16k33_clear
    mode: queued
    then:
      - lambda: |-
          uint8_t clear_cmd[] = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                                 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
          id(i2c_bus).write(0x71, clear_cmd, sizeof(clear_cmd));
          id(i2c_bus).write(0x70, clear_cmd, sizeof(clear_cmd));
          delay(10);







############################# DISPLAY GLUCOSE and TREND ##################################
# Segment map: 14-segment map.xlsx
#
  - id: ht16k33_display_glucose
    mode: queued
    then:
      - script.execute: flash_led
      - lambda: |-
          // Get the glucose value from the sensor
          std::string glucose_value_str = id(glucose_sensor).state.c_str();
          int glucose_value_int = atoi(glucose_value_str.c_str());
          char buffer[5];  // Buffer to hold 3-digit glucose value and null terminator
          snprintf(buffer, sizeof(buffer), "%03d", glucose_value_int);
          ESP_LOGI("DEBUG", "glucose buffer: %s", buffer);

          // Define the segment data for the digits
          static const uint16_t numbers[] = {
            0b0000110000111111, // 0
            0b0000000000000110, // 1
            0b0000000011011011, // 2
            0b0000000011001111, // 3
            0b0000000011100110, // 4
            0b0000000011101101, // 5
            0b0000000011111101, // 6
            0b0000000000000111, // 7
            0b0000000011111111, // 8
            0b0000000011101111  // 9
          };

          // Convert the glucose value string in buffer[] to segment data
          uint16_t digits[] = {
            numbers[buffer[0] - '0'],
            numbers[buffer[1] - '0'],
            numbers[buffer[2] - '0'],
            numbers[4]             // The last digit can be 0 or any static character
          };

          // Remove leading zero
          if (buffer[0] == '0') {
            digits[0] = 0b0000000000000000; // Blank character
          }
          


          // ----- Trend -----
          // Get the trend value from the sensor
          int trend_value = atoi(id(trend_sensor).state.c_str());
          ESP_LOGI("DEBUG", "trend_value: %d", trend_value);

          // Define the segment data for the trends
          static const uint16_t trends[] = {
            0b0000000000000000, // 0 Default
            0b0000001000000001, // 1 Rising Quickly
            0b0000001000000000, // 2 Rising
            0b0000010000000000, // 3 Rising Slightly
            0b0000000010000000, // 4 Steady
            0b0010000000000000, // 5 Falling Slightly
            0b0001000000000000, // 6 Falling
            0b0001000000001000  // 7 Falling Quickly
          };

          // Change digit[3] to the trend indication.
          digits[3] = trends[trend_value];


          // Write the segment data to the HT16K33
          for (int digit = 0; digit < 4; digit++) {
            uint16_t segment_data = digits[digit];
            uint8_t msb = static_cast<uint8_t>(segment_data >> 8);
            uint8_t lsb = static_cast<uint8_t>(segment_data & 0xFF);
            uint8_t write_digit_lsb[] = {static_cast<uint8_t>(digit * 2), lsb};
            uint8_t write_digit_msb[] = {static_cast<uint8_t>(digit * 2 + 1), msb};
            id(i2c_bus).write(0x70, write_digit_lsb, sizeof(write_digit_lsb));
            id(i2c_bus).write(0x70, write_digit_msb, sizeof(write_digit_msb));
            //ESP_LOGI("DEBUG", "Digit: %d, LSB: %u, MSB: %u", digit, lsb, msb);
          }

          // Flash the indicator LED
          //id(flash_led).execute();


################################ DISPLAY TIME #########################################
  - id: ht16k33_display_time
    mode: queued
    then:
      - script.execute: flash_led
      - lambda: |-
          auto time = id(homeassistant_time).now();


          char buffer[9];
          snprintf(buffer, sizeof(buffer), "%02d:%02d", time.hour, time.minute);

          std::string time_str(buffer);
          if (time_str.length() != 5) return;

          // Extract hours and minutes
          int hour = atoi(time_str.substr(0, 2).c_str());
          int minute = atoi(time_str.substr(3, 2).c_str());

          // Make the hour into 12-hour format
            if (hour == 0) {
              hour = 12;  // Midnight case
            } else if (hour > 12) {
              hour -= 12;
            }


          //ESP_LOGI("DEBUG", "hour: %d", hour);
          //ESP_LOGI("DEBUG", "minute: %d", minute);

          // Define the segment data with DP bit for the colon
          static const uint16_t numbers[] = {
            0b0000110000111111, // 0
            0b0000000000000110, // 1
            0b0000000011011011, // 2
            0b0000000011001111, // 3
            0b0000000011100110, // 4
            0b0000000011101101, // 5
            0b0000000011111101, // 6
            0b0000000000000111, // 7
            0b0000000011111111, // 8
            0b0000000011101111  // 9
          };


          // Convert the time to segment data
          uint16_t digits[] = {
            numbers[hour / 10],  // The tens digit of the hour
            static_cast<uint16_t>(numbers[hour % 10] | 0b0100000000000000),  // The units digit of the hour with DP bit set
            numbers[minute / 10],  // The tens digit of the minutes
            numbers[minute % 10]   // The units digit of the minutes
          };


          // Remove leading zero
          if (hour / 10 == 0) {
            digits[0] = 0b0000000000000000; // Blank character
          }
          

          // Write the segment data to the HT16K33
          for (int digit = 0; digit < 4; digit++) {
            uint16_t segment_data = digits[digit];
            uint8_t msb = static_cast<uint8_t>(segment_data >> 8);
            uint8_t lsb = static_cast<uint8_t>(segment_data & 0xFF);
            uint8_t write_digit_lsb[] = {static_cast<uint8_t>(digit * 2), lsb};
            uint8_t write_digit_msb[] = {static_cast<uint8_t>(digit * 2 + 1), msb};
            id(i2c_bus).write(0x71, write_digit_lsb, sizeof(write_digit_lsb));
            id(i2c_bus).write(0x71, write_digit_msb, sizeof(write_digit_msb));
            //ESP_LOGI("DEBUG", "Digit: %d, LSB: %u, MSB: %u", digit, lsb, msb);
          }

################################ TIME TITLE #########################################
# Script to display "TIME" on the HT16K33 display on address 0x71
  - id: ht16k33_display_time_title
    mode: queued
    then:
      - lambda: |-
          // Define the segment data for the title "TIME"
          static const uint16_t chars[] = {
            0b0001001000000001, // T
            0b0001001000001001, // I
            0b0000010100110110, // M
            0b0000000001111001  // E
          };

          // Write the segment data to the HT16K33
          for (int digit = 0; digit < 4; digit++) {
            uint16_t segment_data = chars[digit];
            uint8_t msb = static_cast<uint8_t>(segment_data >> 8);
            uint8_t lsb = static_cast<uint8_t>(segment_data & 0xFF);
            uint8_t write_digit_lsb[] = {static_cast<uint8_t>(digit * 2), lsb};
            uint8_t write_digit_msb[] = {static_cast<uint8_t>(digit * 2 + 1), msb};
            id(i2c_bus).write(0x71, write_digit_lsb, sizeof(write_digit_lsb));
            id(i2c_bus).write(0x71, write_digit_msb, sizeof(write_digit_msb));
          }

################################ BG TITLE #########################################
# Script to display "bg  " on the HT16K33 display on address 0x70
  - id: ht16k33_display_bg_title
    mode: queued
    then:
      - lambda: |-
          // Define the segment data for the title "TIME"
          static const uint16_t chars[] = {
            0b0000000011111100,  //b
            0b0000000010111101,  //g
            0b0000000000000000, // 
            0b0000000000000000  // 
          };


          // Write the segment data to the HT16K33
          for (int digit = 0; digit < 4; digit++) {
            uint16_t segment_data = chars[digit];
            uint8_t msb = static_cast<uint8_t>(segment_data >> 8);
            uint8_t lsb = static_cast<uint8_t>(segment_data & 0xFF);
            uint8_t write_digit_lsb[] = {static_cast<uint8_t>(digit * 2), lsb};
            uint8_t write_digit_msb[] = {static_cast<uint8_t>(digit * 2 + 1), msb};
            id(i2c_bus).write(0x70, write_digit_lsb, sizeof(write_digit_lsb));
            id(i2c_bus).write(0x70, write_digit_msb, sizeof(write_digit_msb));
          }
#####################################################################################

# Script to init and clear the HT16K33 display
  - id: ht16k33_display_startup
    mode: queued
    then:
      - switch.turn_on: data_indicator
      - script.execute: ht16k33_init
      - delay: 500ms
      - script.execute: ht16k33_clear
      - delay: 10ms
      - script.execute: ht16k33_display_time_title
      - script.execute: ht16k33_display_bg_title
      - delay: 2000ms
      - script.execute: ht16k33_display_time
      - switch.turn_off: data_indicator


# Scripts for setting brightness
  - id: set_brightness_low
    mode: queued
    then:
      - lambda: |-
         uint8_t brightness[] = {0xE0};
         id(i2c_bus).write(0x71, brightness, sizeof(brightness));
         delay(10);
         id(i2c_bus).write(0x70, brightness, sizeof(brightness));
         delay(10);

  - id: set_brightness_medium
    mode: queued
    then:
      - lambda: |-
         uint8_t brightness[] = {0xE8};
         id(i2c_bus).write(0x71, brightness, sizeof(brightness));
         delay(10);
         id(i2c_bus).write(0x70, brightness, sizeof(brightness));
         delay(10);


  # Script to flash the data indicator LED
  - id: flash_led
    mode: queued
    then:
      - switch.turn_on: data_indicator
      - delay: 1ms
      - switch.turn_off: data_indicator


# Blue LED, data indicator
switch:
  - platform: gpio
    id: data_indicator
    pin:
      number: D6
      mode: output



`
